---
## Hausaufgabe 5 - Metaheuristiken
---

### Achtung: Bitte überprüfen Sie vor Abgabe der Hausaufgabe, ob das Notebook richtig gespeichert wurde. 
Die Speicherung von Notebooks funktioniert über das "Disketten"-Symbol im Notebook oder über die Shortcuts Strg + S (Win) und CMD + S (MAC).

---
## Bewertung
---

#### Erreichbare Punkte: 25
---
#### Erreichte Punkte:
---
---
Aufgabe 1: <br>
<br>
Aufgabe 2: <br>
<br>
Aufgabe 3: <br>

#### Aufgabe 1 - Zeitmessung (4 Punkte)

Die Performance von Metaheuristiken wird neben der Lösungsqualität auch häufig über die benötigte Laufzeit gemessen. Implementieren Sie eine Zeitmessung für die Durchführung des Solver unter Benutzung des `default_timer` der Bibliothek `timeit`. Der Startpunkt der Messung ist direkt vor `RunLocalSearch` und der Endpunkt nach der Ausgabe der besten gefundenen Lösung. 

Ergänzen Sie den gegebenen Quellcode um die Zeitmessung.


In [1]:
from Solver import *
from timeit import default_timer as timer

data = InputData("../../Prescriptive_Lessons/lesson_5/TestInstancesJson/Small/VFR20_10_1_SIST.json")
solver = Solver(data, 1008)

localSearch = IterativeImprovement(data, 'BestImprovement', ['Insertion'])
iteratedGreedy = IteratedGreedy(
    inputData = data, 
    numberJobsToRemove = 2, 
    baseTemperature = 0.8, 
    maxIterations = 10,
    localSearchAlgorithm = localSearch)

timer_start = timer()
solver.RunLocalSearch(
    constructiveSolutionMethod='NEH',
    algorithm=iteratedGreedy)
timer_end = timer()
print(f'Runtime: {timer_end - timer_start} seconds')

Generating an initial solution according to NEH.
Constructive solution found.
The permutation [14, 5, 12, 10, 17, 8, 4, 11, 2, 15, 1, 18, 0, 16, 13, 19, 7, 3, 9, 6] results in a Makespan of 1665
New best solution in iteration 7: The permutation [14, 12, 8, 10, 2, 13, 18, 16, 0, 1, 11, 19, 7, 3, 6, 9, 17, 5, 15, 4] results in a Makespan of 1569
Best found Solution.
The permutation [14, 12, 8, 10, 2, 13, 18, 16, 0, 1, 11, 19, 7, 3, 6, 9, 17, 5, 15, 4] results in a Makespan of 1569
Runtime: 1.731566600035876 seconds


Mögliches erwartetes Ergebnis (Runtime kann abweichen):

    Generating an initial solution according to NEH.
    Constructive solution found.
    The permutation [14, 5, 12, 10, 17, 8, 4, 11, 2, 15, 1, 18, 0, 16, 13, 19, 7, 3, 9, 6] results in a Makespan of 1665
    New best solution in iteration 7: The permutation [14, 12, 8, 10, 2, 13, 18, 16, 0, 1, 11, 19, 7, 3, 6, 9, 17, 5, 15, 4] results in a Makespan of 1569
    Best found Solution.
    The permutation [14, 12, 8, 10, 2, 13, 18, 16, 0, 1, 11, 19, 7, 3, 6, 9, 17, 5, 15, 4] results in a Makespan of 1569
    Runtime: 1.5346496002748609 seconds

#### Aufgabe 2: Stoppkriterium (6 Punkte)

Neben der gesamten Anzahl an Iterationen wird häufig auch die Anzahl an Iterationen ohne Verbesserung der bisher besten gefunden Lösung als Stoppkriterium gewählt. Ergänzen Sie die Klasse `IteratedGreedy` in `ImprovementAlgorithm.py` so um einen Member `MaxIterationsWithoutImprovement`, dass dessen Wert analog zu `MaxIterations` im Konstruktur übergeben werden kann. 

Die While-Schleife in der Funktion `Run` soll abbrechen, wenn eins der beiden Stoppkriterien erreicht ist.

**Tipp:**
<details>

Neben der Zählvariable `i` wird eine weitere Zählvariable benötigt. Diese wird erhöht, wenn keine beste Lösung gefunden wird und zurückgesetzt, wenn eine neue beste Lösung gefunden wird.
</details>

Ergänzen Sie untenstehenden Quellcode um Ihre neue Variante von Iterated Greedy. Nehmen Sie für Parameter `MaxIterationsWithoutImprovement` den Wert 2 an und für alle anderen die Werte aus Aufgabe 1. 

In [2]:
from Solver import *

data = InputData("../../Prescriptive_Lessons/lesson_5/TestInstancesJson/Small/VFR20_10_1_SIST.json")
solver = Solver(data, 1008)


localSearch = IterativeImprovement(data, 'BestImprovement', ['Insertion'])
iteratedGreedy = IteratedGreedy(
    inputData = data, 
    numberJobsToRemove = 2, 
    baseTemperature = 0.8, 
    maxIterations = 10,
    maxIterationsWithoutImprovement = 2,
    localSearchAlgorithm = localSearch)

solver.RunLocalSearch(
    constructiveSolutionMethod='NEH',
    algorithm=iteratedGreedy)


Generating an initial solution according to NEH.
Constructive solution found.
The permutation [14, 5, 12, 10, 17, 8, 4, 11, 2, 15, 1, 18, 0, 16, 13, 19, 7, 3, 9, 6] results in a Makespan of 1665
New best solution in iteration 7: The permutation [14, 12, 8, 10, 2, 13, 18, 16, 0, 1, 11, 19, 7, 3, 6, 9, 17, 5, 15, 4] results in a Makespan of 1569
Best found Solution.
The permutation [14, 12, 8, 10, 2, 13, 18, 16, 0, 1, 11, 19, 7, 3, 6, 9, 17, 5, 15, 4] results in a Makespan of 1569


Mögliches erwartetes Ergebis:
  
    Generating an initial solution according to NEH.
    Constructive solution found.
    The permutation [14, 5, 12, 10, 17, 8, 4, 11, 2, 15, 1, 18, 0, 16, 13, 19, 7, 3, 9, 6] results in a Makespan of 1665
    Best found Solution.
    The permutation [14, 5, 12, 10, 17, 8, 2, 13, 1, 18, 0, 16, 11, 19, 7, 3, 9, 6, 15, 4] results in a Makespan of 1595

#### Aufgabe 3: Rechenstudie (15 Punkte)

Rechenstudien werden zur Bewertung von Algorithmen und zum Finden guter Parameterkombinationen durchgeführt. Im Folgenden sollen Sie eine kleine Rechenstudie mit der Instanz `VFR20_10_1_SIST.json` durchführen. Testen Sie __alle__ Kombinationen (vollfaktorieller Versuchsplan) der Parameter:

- Lokale Suche mit `BestImprovement`: keine lokale Suche (keine Nachbarschaft), Insertion-Nachbarschaft, TaillardInsertion-Nachbarschaft 
- numberJobsToRemove: 2, 3, 4
- baseTemperature: 0.5, 1
- maxIterations 1, 10
- maxIterationsWithoutImprovement soll keine Rolle spielen (bspw. einfach Wert > maxIterations).



Da Iterated Greedy vom Zufall beeinflusst ist, führen Sie __jeweils__ 3 Iterationen mit unterschiedlichen Seeds für __jede__ Parameterkombinationen durch. Mit `iteration` als Laufvariable für die aktuelle Iteration, soll der Seed wie folgt berechnet werden:
seed = `numberJobsToRemove`  x  `maxIterations`  x  `iteration`

Speichern Sie für jede Iteration die folgenden Daten in einem Dictionary (https://docs.python.org/3/tutorial/datastructures.html#dictionaries):

- Iteration (Key: Iteration)
- Jeden Parameter, außer maxIterationsWithoutImprovement (siehe Aufzählung oben) als einzelner Eintrag (Keys: LocalSearch, NumberJobsToRemove, BaseTemperature, MaxIterations)
- Seed (Key: Seed)
- Makespan(Key: Makespan)
- Laufzeit (Key: Runtime; vgl. Aufgabe 1)

Fügen Sie dann das Dictionary einer Liste `rows` hinzu. Erzeugen Sie nach dem alle Versuche beendet sind aus `rows` einen pandas data frame `results` (https://stackoverflow.com/a/17496530). Bilden Sie mit der Funktion `groupby` den Mittelwert (https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.GroupBy.mean.html) von `Makespan` und `Runtime` für die 4 Parameter `LocalSearch`, `NumberJobsToRemove`, `BaseTemperature`, `MaxIterations`.



In [12]:
from timeit import default_timer as timer
import pandas as pd

neighborhoodList = [[], ['Insertion'], ['TaillardInsertion']]
numberJobsToRemoveList = [2, 3, 4]
baseTemperatureList = [0.5, 1]
maxIterationsList = [1, 10]
maxIterationsWithoutImprovement = 15

iterations = 3

rows = []

data = InputData("../../Prescriptive_Lessons/lesson_5/TestInstancesJson/Small/VFR20_10_1_SIST.json")

for neighborhood in neighborhoodList:
    for number in numberJobsToRemoveList:
        for temperature in baseTemperatureList:
            for maxIter in maxIterationsList:
                for iteration in range(iterations):
                    seed = number*maxIter*iteration
                    localSearch = IterativeImprovement(data, 'BestImprovement', neighborhood)
                    solver = Solver(data, seed)
                    iteratedGreedy = IteratedGreedy(
                        inputData = data,
                        numberJobsToRemove = number,
                        baseTemperature = temperature,
                        maxIterations = maxIter,
                        maxIterationsWithoutImprovement = maxIter+1,
                        localSearchAlgorithm = localSearch)
                    
                    start = timer()
                    solver.RunLocalSearch(
                        constructiveSolutionMethod='NEH',
                        algorithm=iteratedGreedy)
                    end = timer()

                    if neighborhood:
                        a = neighborhood[0]
                    else:
                        a = 'None'

                    row = {
                        'Iteration': iteration,
                        'LocalSearch': a,
                        'NumberJobsToRemove': number,
                        'BaseTemperature': temperature,
                        'MaxIterations': maxIter,
                        'Seed': seed,
                        'Makespan': solver.SolutionPool.GetLowestMakespanSolution().Makespan,
                        'Runtime': end-start
                    }

                    rows.append(row)

Generating an initial solution according to NEH.
Constructive solution found.
The permutation [14, 5, 12, 10, 17, 8, 4, 11, 2, 15, 1, 18, 0, 16, 13, 19, 7, 3, 9, 6] results in a Makespan of 1665
Best found Solution.
The permutation [14, 5, 12, 10, 17, 8, 4, 11, 2, 15, 1, 18, 0, 16, 13, 19, 7, 3, 9, 6] results in a Makespan of 1665
Generating an initial solution according to NEH.
Constructive solution found.
The permutation [14, 5, 12, 10, 17, 8, 4, 11, 2, 15, 1, 18, 0, 16, 13, 19, 7, 3, 9, 6] results in a Makespan of 1665
Best found Solution.
The permutation [14, 5, 12, 10, 17, 8, 4, 11, 2, 15, 1, 18, 0, 16, 13, 19, 7, 3, 9, 6] results in a Makespan of 1665
Generating an initial solution according to NEH.
Constructive solution found.
The permutation [14, 5, 12, 10, 17, 8, 4, 11, 2, 15, 1, 18, 0, 16, 13, 19, 7, 3, 9, 6] results in a Makespan of 1665
Best found Solution.
The permutation [14, 5, 12, 10, 17, 8, 4, 11, 2, 15, 1, 18, 0, 16, 13, 19, 7, 3, 9, 6] results in a Makespan of 1665
G

In [13]:
results = pd.DataFrame(rows)
print(results)

groupedResults = results.groupby(['LocalSearch', 'NumberJobsToRemove', 'BaseTemperature', 'MaxIterations'])[['Makespan', 'Runtime']].mean()
print(groupedResults)

     Iteration        LocalSearch  NumberJobsToRemove  BaseTemperature  \
0            0               None                   2              0.5   
1            1               None                   2              0.5   
2            2               None                   2              0.5   
3            0               None                   2              0.5   
4            1               None                   2              0.5   
..         ...                ...                 ...              ...   
103          1  TaillardInsertion                   4              1.0   
104          2  TaillardInsertion                   4              1.0   
105          0  TaillardInsertion                   4              1.0   
106          1  TaillardInsertion                   4              1.0   
107          2  TaillardInsertion                   4              1.0   

     MaxIterations  Seed  Makespan   Runtime  
0                1     0      1665  0.003241  
1                

In [ ]:
bestMakespan = results['Makespan'].min()
results[results['Makespan'] == bestMakespan]
results.Makespan.unique()

array([1665, 1628, 1633, 1598, 1649, 1664, 1607, 1608, 1635, 1612, 1604,
       1622, 1619, 1595, 1568, 1569, 1556, 1553, 1551, 1571, 1586, 1583,
       1548, 1561, 1542])

Mögliches erwartetes Ergebnis in der Ausgabe:

![Ausgabe](Ausgabe.PNG)